File: hackerrank/problem solving/algorithms/2 - implementation/46 - beautiful triplets/oo/beautiful_triplets_oo.py
# Source: https://www.hackerrank.com/challenges/beautiful-triplets/problem?isFullScreen=true

def main():

    SIZE, BEAUTIFUL_DIFFERENCE = read_an_array()
    ARRAY = read_an_array()

    OBJ = BeatifulTriplets(ARRAY, BEAUTIFUL_DIFFERENCE)
    print(OBJ.get_n_beautiful_triplets())


def read_an_array():

    ARRAY = list(map(int, input().split()))
    return ARRAY


class BeatifulTriplets:

    def __init__(self, array, beautiful_difference):

        self._array = array
        self._beautiful_difference = beautiful_difference
        self._find_beautiful_triplets()


    def _find_beautiful_triplets(self):

        self._n_beautiful_triplets = 0

        if len(self._array) > 2:
            for i in range(0, len(self._array)-2):
                for j in range(i+1, len(self._array)-1):
                    FIRST_DIFFERENCE = self._array[j] - self._array[i]

                    if FIRST_DIFFERENCE < self._beautiful_difference:
                        continue
                    elif FIRST_DIFFERENCE > self._beautiful_difference:
                        break
                    
                    for k in range(j+1, len(self._array)):
                        SECOND_DIFFERENCE = self._array[k] - self._array[j]

                        if SECOND_DIFFERENCE == self._beautiful_difference:
                            self._n_beautiful_triplets += 1
                        elif SECOND_DIFFERENCE > self._beautiful_difference:
                            break


    def get_n_beautiful_triplets(self):

        return self._n_beautiful_triplets


if __name__ == "__main__":
    main()

File: hackerrank/problem solving/algorithms/2 - implementation/46 - beautiful triplets/oo/beautiful_triplets_oo.cs
// Source: https://www.hackerrank.com/challenges/beautiful-triplets/problem?isFullScreen=true

using System;

public class Solution
{
    public static void Main()
    {
        List<int> inputLine = _readAnArray();
        int size = inputLine.First();
        int beautifulDifference = inputLine.Last();
        List<int> array = _readAnArray();

        BeautifulTriplets obj = new BeautifulTriplets(array, beautifulDifference);
        Console.WriteLine(obj.GetNBeautifulTriplets());
    }

        private static List<int> _readAnArray()
        {
            List<int> array = Console.ReadLine().Split().Select(int.Parse).ToList();
            return array;
        }
}

    public class BeautifulTriplets
    {
        private List<int> _array;
        private int _beautifulDifference;
        private int _nBeautifulTriplets;

        public BeautifulTriplets(List<int> array, int beautifulDifference)
        {
            _array = array;
            _beautifulDifference = beautifulDifference;
            _findBeautifulTriplets();
        }

            private void _findBeautifulTriplets()
            {
                _nBeautifulTriplets = 0;

                if (_array.Count > 2)
                    for (int i = 0; i < _array.Count - 2; i++)
                        for (int j = i + 1; j < _array.Count - 1; j++)
                        {
                            int firstDifference = _array[j] - _array[i];

                            if (firstDifference < _beautifulDifference)
                                continue;
                            else if (firstDifference > _beautifulDifference)
                                break;

                            for (int k = j + 1; k < _array.Count; k++)
                            {
                                int secondDifference = _array[k] - _array[j];

                                if (secondDifference == _beautifulDifference)
                                    _nBeautifulTriplets++;
                                else if (secondDifference > _beautifulDifference)
                                    break;
                            }
                        }
            }

            public int GetNBeautifulTriplets()
            {
                return _nBeautifulTriplets;
            }
        }
